xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/objects/classes/DBListClass.java

```diff
@@ -282,8 +285,16 @@ public String getQuery(XWikiContext context)
                     select.append("doc." + idField);
                 } else {
                     select.append("idprop.value");
-                    fromStatements.add("StringProperty as idprop");
                     whereStatements.add("obj.id=idprop.id.id and idprop.id.name='" + idField + "'");
+                    // Get the from statements according to the type of property
+                    try {
+                        PropertyClass pc = (PropertyClass) context.getWiki().getDocument(classname, context).getXClass()
+                                .get(idField);
+                        String classType = (pc == null) ? "StringProperty" : pc.newProperty().getClass().getName();
+                        fromStatements.add(classType + " as idprop");
+                    } catch (XWikiException e) {
+                        e.printStackTrace();
```

Comment:
Exceptions need to be handled properly. The severity needs to be analyzed and a proper solution found. If the error is not important, then logging can be used (See http://dev.xwiki.org/xwiki/bin/view/Community/JavaCodeStyle#HLoggingBestPractices). OTOH if the error is important and we cannot continue then an exception needs to be raised.